mvrx/src/main/kotlin/com/airbnb/mvrx/MvCorStateStore.kt

```diff
@@ -0,0 +1,117 @@
+package com.airbnb.mvrx
+
+import io.reactivex.Observable
+import io.reactivex.disposables.Disposable
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
+import kotlinx.coroutines.channels.actor
+import kotlinx.coroutines.channels.consumeEach
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.rx2.asObservable
+import java.util.*
+import kotlin.coroutines.CoroutineContext
+
+
+/**
+ * This is a container class around the actual state itself. It has a few optimizations to ensure
+ * safe usage of state.
+ *
+ * All state reducers are run in an actor to ensure that they don't have race
+ * conditions with each other.
+ *
+ */
+
+internal open class MvCorStateStore<S : Any>(initialState: S, final override val coroutineContext: CoroutineContext = Dispatchers.Default) : Disposable, CoroutineScope, MvRxStateStore<S> {
+
+    /**
+     * The stateChannel is where state changes should be pushed to.
+     */
+    private val stateChannel = ConflatedBroadcastChannel(initialState)
+    /**
+     * The observable observes the stateChannel but only emits events when the state actually changed.
+     */
+
+    override val observable: Observable<S> = stateChannel.openSubscription().asObservable(coroutineContext).distinctUntilChanged()
+    /**
+     * This is automatically updated from a subscription on the stateChannel for easy access to the
+     * current state.
+     */
+    override val state: S
+        get() = stateChannel.value
+
+    sealed class Job<S> {
+        class SetQueueElement<S>(val reducer: S.() -> S) : Job<S>()
+        class GetQueueElement<S>(val block: (S) -> Unit) : Job<S>()
+    }
+
+    /**
+     * Actor responsible for processing get and set blocks. Sequentially processes every block, SetQueueElement in higher priority.
+     * Once every SetQueueElement is processed actor iterates over GetQueueElements until no elements are left, or a new message
+     * is sent to an actor
+     */
+    private val actor = actor<Job<S>>(coroutineContext, capacity = Channel.UNLIMITED) {
```

Comment:
I'm not perfectly primed on actors yet, what is the advantage of using an actor rather than a straight channel or flow here?

Reply:
It's basically a shorhand for this:
```
private val queueChannel = Channel<Job<S>>(capacity = Channel.UNLIMITED)
private val actor = launch {
   queueChannel.consumeEach {
```

From documentation:
An actor is an entity made up of a combination of a coroutine, the state that is confined and encapsulated into this coroutine, and a channel to communicate with other coroutines. A simple actor can be written as a function, but an actor with a complex state is better suited for a class.

There is an actor coroutine builder that conveniently combines actor's mailbox channel into its scope to receive messages from and combines the send channel into the resulting job object, so that a single reference to the actor can be carried around as its handle.